torch_geometric/nn/conv/gps_conv.py

```diff
@@ -159,4 +181,180 @@ def forward(
 
     def __repr__(self) -> str:
         return (f'{self.__class__.__name__}({self.channels}, '
-                f'conv={self.conv}, heads={self.heads})')
+                f'conv={self.conv}, heads={self.heads}, '
+                f'attn_type={self.attn_type})')
+
+
+def _orthogonal_matrix(dim: int) -> Tensor:
+    r"""Get an orthogonal matrix by applying QR decomposition."""
+    # Random matrix from normal distribution
+    mat = torch.randn((dim, dim))
+    # QR decomposition to two orthogonal matrices
+    q, _ = torch.linalg.qr(mat.cpu(), mode='reduced')
+    return q.t()
+
+
+def orthogonal_matrix(num_rows: int, num_cols: int) -> Tensor:
+    r"""Generate an orthogonal matrix with `num_rows` rows
+    and `num_cols` columns.
+    """
+    num_full_blocks = int(num_rows / num_cols)
+    blocks = []
+    for _ in range(num_full_blocks):
+        q = _orthogonal_matrix(num_cols)
+        blocks.append(q)
+    remain_rows = num_rows - num_full_blocks * num_cols
+    if remain_rows > 0:
+        q = _orthogonal_matrix(num_cols)
+        blocks.append(q[:remain_rows])
+    mat = torch.cat(blocks)
+    # multiplier = torch.randn((num_rows, num_cols)).norm(dim=1)
+    # scaler = torch.diag(multiplier)
+    # mat = scaler @ mat
+    return mat
+
+
+def linear_attention(q: Tensor, k: Tensor, v: Tensor) -> Tensor:
+    r"""Efficient attention mechanism from the
+    `"Rethinking Attention with Performers"
+    <https://arxiv.org/abs/2009.14794>`_ paper.
+
+    .. math::
+        \mathbf{\hat{D}}^{-1}(\mathbf{Q}'((\mathbf{K}')^{\top} \mathbf{V}))
+
+    """
+    k_contract = k.sum(dim=-2)
+    D_inv = 1.0 / torch.einsum('...Lr,...r->...L', q, k_contract)
+    kv = torch.einsum('...Lr,...Ld->...rd', k, v)
+    qkv = torch.einsum('...Lr,...rd->...Ld', q, kv)
+    out = torch.einsum('...L,...Ld->...Ld', D_inv, qkv)
+    return out
+
+
+def generalized_kernel(x: Tensor, mat: Tensor,
+                       kernel: Callable = torch.nn.ReLU(),
+                       epsilon: float = 0.001) -> Tensor:
+    r"""Apply generalized kernelizable attention with
+    kernel functions such as the ReLU.
+    """
+    num_batches, num_heads, *_ = x.shape
+    # Expand projection matrix to number of batches and number of heads
+    projection = mat.expand(num_batches, num_heads, *mat.shape)
+    # "Inner" product x with projection matrix
+    x = torch.einsum('...id,...jd->...ij', x, projection)
+    out = kernel(x) + epsilon
+    return out
+
+
+class PerformerFastAttention(torch.nn.Module):
```

Comment:
Can we rename this . `PerformerFastAttention` seems like a faster version of `PerformerAttention` :).